### DISCUSSION TITLE
Error: render function 'get_build_state' in template 'index' failed (cause: Server(None))

### DISCUSSION THREAD

**Discussion by afidegnum** [framesurge/perseus#143] [Q&A] [answered]
Hi, 

Can you please give a hint on fixing the error, I tried unsuccessful debugging. 

```

Error: render function 'get_build_state' in template 'index' failed (cause: Server(None))

Caused by:
        Network Error: Failed to read JSON: invalid type: map, expected i32 at line 1 column 2
        Caused by:
                Failed to read JSON: invalid type: map, expected i32 at line 1 column 2
```
Here is the code:
The API supposed to display a list of categories. 

```rs
use perseus::{Html, RenderFnResultWithCause, Template};
use serde::{Deserialize, Serialize};
use sycamore::prelude::*;

#[perseus::make_rx(IndexPageStateRx)]
pub struct IndexPageState {
    server_ip: String,
    category: Vec<Category>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MarkdownPageProps {
    pub text: String,
}

#[derive(Serialize, Clone, Deserialize, PartialEq, Eq)]
pub struct Category {
    pub id: i32,
    pub name: String,
    pub slug: String,
    pub description: String,
}

#[perseus::template_rx]
pub fn index_page(state: IndexPageStateRx) -> View<G> {
    let sever_ip = state.server_ip.handle();
    let category = state.category.handle();

    view! {
        body  {

            ul {
             Keyed(KeyedProps {
                 iterable: category,

                 //  template: cloned!(x => move |x| view! {
                 //  li { (format!("{:?}", x.title)) }
                 // }),

                 template: |x| view! {
                     li { ( format!("{:?}", x.name)) }
                 },
                 key: |x| x.id,
             })
         }

            ul (data-table="menu", class="menu") {div (class="itm") {"home"}


                                                  div (class="itm") {"About us"}

                                                  div (class="itm") {"Contact us"}

            }

                div (id="innerone") {h1  {"This Title Update"}
                }

                div (id="innertwo") {h2  {"Subheads"}
                }

            }

        }

    }
}

pub fn get_template<G: Html>() -> Template<G> {
    Template::new("index")
        .build_state_fn(get_build_state)
        .template(index_page)
}

#[perseus::autoserde(build_state)]
pub async fn get_build_state(
    _path: String,
    _locale: String,
) -> RenderFnResultWithCause<IndexPageState> {
    //We'll cache the result with `try_cache_res`, which means we only make the request once, and future builds will use the cached result (speeds up development)
    let body = perseus::cache_fallible_res(
        "ipify",
        || async {
            let res: Category = ureq::get("http://localhost:8081/categories/")
                .call()?
                .into_json()?;
            Ok::<Vec<Category>, ureq::Error>(res)
        },
        false,
    )
    .await?;

    Ok(IndexPageState {
        server_ip: String::from(""),
        category: vec![body],
    })
}
```

**Comment by afidegnum** [comment, marked as answer]
I found it, I needed to change the types. to Vec<Category>

**Reply by arctic-hen7** [reply]
I might suggest that these kinds of questions are better on the Discord server in future.

**Comment by afidegnum** [comment]
ok, there is an update on the rendering. I just posted it on Discord

**Comment by afidegnum** [comment]
The views disappears right after loading. 
when using `ureq`
from the above code. please see the attached video.


https://user-images.githubusercontent.com/374206/167841902-97dc862e-a550-45cc-b8d7-e7cb3242f3c9.mp4

**Reply by afidegnum** [reply]
Followed by this error:
```js
bundle.js:998 Uncaught (in promise) Error: `unwrap_throw` failed
    at imports.wbg.__wbindgen_throw (bundle.js:998:11)
    at wasm_bindgen::throw_str::h7d08a954e2b7118d (bundle.wasm:0x24ac4b)
    at <core::result::Result<T,E> as wasm_bindgen::UnwrapThrowExt<T>>::expect_throw::h073b7eb0af8f9797 (bundle.wasm:0x18d18b)
    at wasm_bindgen::UnwrapThrowExt::unwrap_throw::hfd4b2fc8b1bd7e43 (bundle.wasm:0x21e451)
    at <sycamore::generic_node::dom_node::DomNode as sycamore::generic_node::GenericNode>::insert_child_before::he92079b9a0834893 (bundle.wasm:0x1a697d)
    at <sycamore::generic_node::hydrate_dom::HydrateNode as sycamore::generic_node::GenericNode>::insert_child_before::h8ea15163cf4c858b (bundle.wasm:0x22a1da)
    at sycamore::utils::render::insert_expression::h001669af4b1de9cc (bundle.wasm:0x3240e)
    at sycamore::utils::render::insert::hbdc730d987a18322 (bundle.wasm:0x18ee37)
    at sycamore::generic_node::hydrate_dom::hydrate_get_scope::{{closure}}::h65e14ab4f1ce0e2a (bundle.wasm:0x131658)
    at core::ops::function::FnOnce::call_once{{vtable.shim}}::h9f074b74f9be8ea9 (bundle.wasm:0x2543f3)
```

<html>
<body>


imports.wbg.__wbindgen_throw | @ | bundle.js:998
-- | -- | --
  | $wasm_bindgen::throw_str::h7d08a954e2b7118d | @ | bundle.wasm:0x24ac4b
  | $<core::result::Result<T,E> as wasm_bindgen::UnwrapThrowExt<T>>::expect_throw::h073b7eb0af8f9797 | @ | bundle.wasm:0x18d18b
  | $wasm_bindgen::UnwrapThrowExt::unwrap_throw::hfd4b2fc8b1bd7e43 | @ | bundle.wasm:0x21e451
  | $<sycamore::generic_node::dom_node::DomNode as sycamore::generic_node::GenericNode>::insert_child_before::he92079b9a0834893 | @ | bundle.wasm:0x1a697d
  | $<sycamore::generic_node::hydrate_dom::HydrateNode as sycamore::generic_node::GenericNode>::insert_child_before::h8ea15163cf4c858b | @ | bundle.wasm:0x22a1da
  | $sycamore::utils::render::insert_expression::h001669af4b1de9cc | @ | bundle.wasm:0x3240e
  | $sycamore::utils::render::insert::hbdc730d987a18322 | @ | bundle.wasm:0x18ee37
  | $sycamore::generic_node::hydrate_dom::hydrate_get_scope::{{closure}}::h65e14ab4f1ce0e2a | @ | bundle.wasm:0x131658
  | $core::ops::function::FnOnce::call_once{{vtable.shim}}::h9f074b74f9be8ea9 | @ | bundle.wasm:0x2543f3
  | $<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h56d741e398900b66 | @ | bundle.wasm:0x22fef7
  | $sycamore_reactive::_create_child_scope_in::{{closure}}::hbb27a8e18526a96d | @ | bundle.wasm:0xa56ca
  | $std::thread::local::LocalKey<T>::try_with::h8a62d1ff9e95cc9c | @ | bundle.wasm:0x148c0f
  | $std::thread::local::LocalKey<T>::with::hc1d3b2df04b6d062 | @ | bundle.wasm:0x19c5f3
  | $sycamore_reactive::_create_child_scope_in::h3f7231e85313f27d | @ | bundle.wasm:0x1b17c8
  | $sycamore_reactive::create_scope::h29be3660b3d77498 | @ | bundle.wasm:0x187150
  | $sycamore::generic_node::hydrate_dom::hydrate_get_scope::hbf8b08d2981ee82c | @ | bundle.wasm:0x1af743
  | $sycamore::generic_node::hydrate_dom::hydrate_to::ha0c22be6a8e0c6db | @ | bundle.wasm:0x1beee5
  | $perseus::shell::app_shell::{{closure}}::h2de62d55fe8968a0 | @ | bundle.wasm:0x7d9a
  | $<core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h5e5d5b6f1a5017e5 | @ | bundle.wasm:0x1927b9
  | $perseus::router::router_component::on_route_change::{{closure}}::h20b7e0ccf65a9fa0 | @ | bundle.wasm:0x389ab
  | $<core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h11e58eee231ab751 | @ | bundle.wasm:0x19263b
  | $wasm_bindgen_futures::task::singlethread::Task::run::h0bb3f2e39eb98612 | @ | bundle.wasm:0xdf85f
  | $wasm_bindgen_futures::queue::QueueState::run_all::h9c69f09659ebf257 | @ | bundle.wasm:0xbeb39
  | $wasm_bindgen_futures::queue::Queue::new::{{closure}}::h489ceb750e9b7619 | @ | bundle.wasm:0x1fd575
  | $<dyn core::ops::function::FnMut<(A,)>+Output = R as wasm_bindgen::closure::WasmClosure>::describe::invoke::hc163ee6dd0c8ea26 | @ | bundle.wasm:0x17c752
  | __wbg_adapter_36 | @ | bundle.js:283
  | real | @ | bundle.js:205
  | Promise.then (async) |   |  
  | (anonymous) | @ | bundle.js:954
  | logError | @ | bundle.js:222
  | imports.wbg.__wbg_then_ce526c837d07b68f | @ | bundle.js:953
  | $js_sys::Promise::then::h3f2017f501e3b68f | @ | bundle.wasm:0x1f6e92
  | $wasm_bindgen_futures::queue::Queue::schedule_task::h1cc03a0fb38a3a0a | @ | bundle.wasm:0x11cc04
  | $wasm_bindgen_futures::task::singlethread::Task::spawn::{{closure}}::hd5c6583fd1d69c44 | @ | bundle.wasm:0x24a482
  | $std::thread::local::LocalKey<T>::try_with::h801edf22e12aea78 | @ | bundle.wasm:0x159af4
  | $std::thread::local::LocalKey<T>::with::h430c8636b04b3ec4 | @ | bundle.wasm:0x1f113d
  | $wasm_bindgen_futures::task::singlethread::Task::spawn::h2ce9ebf037e1c690 | @ | bundle.wasm:0xb2847
  | $wasm_bindgen_futures::spawn_local::hbe6d1ccb95949fdc | @ | bundle.wasm:0x1b7587
  | $perseus::router::router_component::on_route_change::h8747c91646d2757e | @ | bundle.wasm:0xc85e5
  | $<perseus::router::router_component::PerseusRouter<G,AppRoute> as sycamore::component::Component<G>>::create_component::{{closure}}::{{closure}}::h37e14fdc4762627a | @ | bundle.wasm:0xf46b6
  | $<alloc::boxed::Box<F,A> as core::ops::function::FnMut<Args>>::call_mut::hc15340a1b7ca6bfc | @ | bundle.wasm:0x22ff49
  | $sycamore_reactive::effect::_create_effect::{{closure}}::{{closure}}::{{closure}}::h2f016032dac3476e | @ | bundle.wasm:0x25a36f
  | $core::ops::function::FnOnce::call_once{{vtable.shim}}::h90a9e5fba8ce1148 | @ | bundle.wasm:0x24b3ab
  | $<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h56d741e398900b66 | @ | bundle.wasm:0x22fef7
  | $sycamore_reactive::_create_child_scope_in::{{closure}}::hbb27a8e18526a96d | @ | bundle.wasm:0xa56ca
  | $std::thread::local::LocalKey<T>::try_with::h8a62d1ff9e95cc9c | @ | bundle.wasm:0x148c0f
  | $std::thread::local::LocalKey<T>::with::hc1d3b2df04b6d062 | @ | bundle.wasm:0x19c5f3
  | $sycamore_reactive::_create_child_scope_in::h3f7231e85313f27d | @ | bundle.wasm:0x1b17c8
  | $sycamore_reactive::create_child_scope_in::h15c7ddfbbaa6d6b7 | @ | bundle.wasm:0x1f6297
  | $sycamore_reactive::effect::_create_effect::{{closure}}::{{closure}}::hd2044e34b691f9fa | @ | bundle.wasm:0x47642
  | $std::thread::local::LocalKey<T>::try_with::h6ff08e1b42a49ceb | @ | bundle.wasm:0x1663a5
  | $std::thread::local::LocalKey<T>::with::h2f6dab46c028a7dc | @ | bundle.wasm:0x1e1e34
  | $sycamore_reactive::effect::_create_effect::{{closure}}::hc2b7cbb6af4d3f40 | @ | bundle.wasm:0x1f6305
  | $sycamore_reactive::effect::_create_effect::h65e73b66641cb710 | @ | bundle.wasm:0x5c246
  | $sycamore_reactive::effect::create_effect::hfa15b77a2234ce59 | @ | bundle.wasm:0xd75fe
  | $<perseus::router::router_component::PerseusRouter<G,AppRoute> as sycamore::component::Component<G>>::create_component::{{closure}}::h86703964a87d94d2 | @ | bundle.wasm:0x5a8d7
  | $<sycamore_router::router::Router<G,R,F> as sycamore::component::Component<G>>::create_component::h1578d9d51d96e93c | @ | bundle.wasm:0x5035a
  | $sycamore::component::instantiate_component::{{closure}}::h9be1f14d3661ed5d | @ | bundle.wasm:0x1ff184
  | $sycamore_reactive::effect::untrack::{{closure}}::h8b6bc1792f45b847 | @ | bundle.wasm:0x10296c
  | $std::thread::local::LocalKey<T>::try_with::h5f20b498b1d28499 | @ | bundle.wasm:0x153cb8
  | $sycamore_reactive::effect::untrack::h25d555ef75599a7e | @ | bundle.wasm:0xaea27
  | $<perseus::router::router_component::PerseusRouter<G,AppRoute> as sycamore::component::Component<G>>::create_component::h1254f4fc06fc21a0 | @ | bundle.wasm:0x48e1e
  | $sycamore::component::instantiate_component::{{closure}}::h2088d6014fd1d869 | @ | bundle.wasm:0x1ff126
  | $sycamore_reactive::effect::untrack::{{closure}}::he7a253808228f977 | @ | bundle.wasm:0x102e64
  | $std::thread::local::LocalKey<T>::try_with::hf4e10e56050f8b03 | @ | bundle.wasm:0x1541e5
  | $sycamore_reactive::effect::untrack::hfab3f361972ce750 | @ | bundle.wasm:0xaed5a
  | $perseus_engine::run::{{closure}}::hbd32545419ce56b2 | @ | bundle.wasm:0x1870a6
  | $sycamore::generic_node::dom_node::render_get_scope::{{closure}}::h384539f0ccfe5442 | @ | bundle.wasm:0x131399
  | $core::ops::function::FnOnce::call_once{{vtable.shim}}::hd28e23cae2775b86 | @ | bundle.wasm:0x2544f7
  | $<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h56d741e398900b66 | @ | bundle.wasm:0x22fef7
  | $sycamore_reactive::_create_child_scope_in::{{closure}}::hbb27a8e18526a96d | @ | bundle.wasm:0xa56ca
  | $std::thread::local::LocalKey<T>::try_with::h8a62d1ff9e95cc9c | @ | bundle.wasm:0x148c0f
  | $std::thread::local::LocalKey<T>::with::hc1d3b2df04b6d062 | @ | bundle.wasm:0x19c5f3
  | $sycamore_reactive::_create_child_scope_in::h3f7231e85313f27d | @ | bundle.wasm:0x1b17c8
  | $sycamore_reactive::create_scope::h509d6b2cd77edde9 | @ | bundle.wasm:0x187212
  | $sycamore::generic_node::dom_node::render_get_scope::h2b3b84c08d37fb37 | @ | bundle.wasm:0x1bee7b
  | $sycamore::generic_node::dom_node::render_to::hb3ae8c7914186bf9 | @ | bundle.wasm:0x1bedcb
  | $perseus_engine::run::h2704afd33f0fc010 | @ | bundle.wasm:0x6a2e2
  | $run | @ | bundle.wasm:0x1cacc9
  | run | @ | bundle.js:298
  | main | @ | (index):14
  | await in main (async) |   |  
  | (anonymous)


</body>
</html>

**Reply by lukechu10** [reply]
Perseus 0.3 is using sycamore 0.7 which has quite a few hydration bugs. Perseus 0.4 should fix these. For now, I would recommend just disabling hydration by removing the `hydrate` feature from perseus. Hope this helps :)

**Reply by afidegnum** [reply]
Ok, Thanks a lot for the tip!

**Reply by afidegnum** [reply]
@arctic-hen7 is it possible to pass SSR node from the head to the body as well? referring from [this example](https://github.com/arctic-hen7/perseus/blob/main/examples/core/router_state/src/templates/about.rs#L12)

**Reply by arctic-hen7** [reply]
Sorry, I don't know what you mean...

Anything in that `head()` function will be automatically rendered on the server though.
